src/batch-transaction.js

```diff
@@ -0,0 +1,343 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+var common = require('@google-cloud/common');
+var extend = require('extend');
+var is = require('is');
+var util = require('util');
+
+var codec = require('./codec.js');
+var Transaction = require('./transaction.js');
+
+/**
+ * Use a BatchTransaction object to create partitions and read/query against
+ * your Cloud Spanner database.
+ *
+ * @class
+ * @extends Transaction
+ *
+ * @param {TransactionOptions} [options] [Transaction options](https://cloud.google.com/spanner/docs/timestamp-bounds).
+ *
+ * @example <caption>include:samples/batch.js</caption>
+ * region_tag:create_batch_transaction
+ */
+function BatchTransaction(session) {
+  Transaction.call(this, session, {readOnly: true});
+}
+
+util.inherits(BatchTransaction, Transaction);
+
+/**
+ * Closes all open resources.
+ *
+ * When the transaction is no longer needed, you should call this method to free
+ * up resources allocated by the Batch client.
+ *
+ * Calling this method would render the transaction unusable everywhere. In
+ * particular if this transaction object was being used across multiple
+ * machines, calling this method on any of the machine would make the
+ * transaction unusable on all the machines. This should only be called when the
+ * transaction is no longer needed anywhere
+ *
+ * @param {BasicCallback} [callback] Callback function.
+ * @returns {Promise<BasicResponse>}
+ *
+ * @example
+ * const Spanner = require('@google-cloud/spanner');
+ * const spanner = new Spanner();
+ *
+ * const instance = spanner.instance('my-instance');
+ * const database = instance.database('my-database');
+ *
+ * database.createBatchTransaction(function(err, transaction) {
+ *   if (err) {
+ *     // Error handling omitted.
+ *   }
+ *
+ *   transaction.close(function(err, apiResponse) {});
+ * });
+ *
+ * //-
+ * // If the callback is omitted, we'll return a Promise.
+ * //-
+ * database.createBatchTransaction().then(function(data) {
+ *   var transaction = data[0];
+ *   return transaction.close();
+ * });
+ */
+BatchTransaction.prototype.close = function(callback) {
+  this.session.delete(callback);
+};
+
+/**
+ * @see [`ExecuteSqlRequest`](https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.v1#google.spanner.v1.ExecuteSqlRequest)
+ * @typedef {object} QueryPartition
+ * @property {string} partitionToken The partition token.
+ */
+/**
+ * @typedef {array} CreateQueryPartitionsResponse
+ * @property {QueryPartition[]} 0 List of query partitions.
+ * @property {object} 1 The full API response.
+ */
+/**
+ * @callback CreateQueryPartitionsCallback
+ * @param {?Error} err Request error, if any.
+ * @param {QueryPartition[]} partitions List of query partitions.
+ * @param {object} apiResponse The full API response.
```

Comment:
No apiResponse from GAPIC here.

Reply:
This isn't a paginated list, so there is a response.

Reply:
Oops, then the way it is currently, the user callback receives what GAPIC returns which is:

https://github.com/callmehiphop/nodejs-spanner/blob/df0116e5f450cc8bf4759e49494b08b92dc83634/src/v1/spanner_client.js#L1202

> The function which will be called with the result of the API call.
> The second parameter to the callback is an object representing [PartitionResponse]{@link google.spanner.v1.PartitionResponse}.

Reply:
*/mark as resolved*